File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null || head.next == null){
            return false;
        }

        ListNode leading = head;
        ListNode trailing = head;

        while(leading != null && leading.next != null){
            leading = leading.next.next;
            trailing = trailing.next;
            if(leading == trailing){
                return true;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        start = head
        l = head
        r = head
        while n > 0:
            r = r.next
            n -= 1
        
        if r is None:
            return head.next

        prev = None
        while r:
            prev = l
            l = l.next
            r = r.next
        
        prev.next = l.next
        return start

        
